### DISCUSSION TITLE
3D symmetry to full 3D solution

### DISCUSSION THREAD

**Discussion by am-tc01** [idaholab/moose#20209] [Q&A Modules: General] [answered]
Hi,
I am trying to convert a 3D symmetric solution to full 3D, read it in a sub-app and Transfer a variable to a main app. 
My first idea was to simply transform the symmetric solution by changing z coordinate to -z coordinate, so that I get a mirrored version of the original solution, read them both using `FileMeshGenerator `and stitch them using `StitchedMeshGenerator`. This didn't work since mirroring probably distorted node order and I got an error of "`Negative Jacobian`". Although not sure if that is the reason, but got this error even though I am not solving for anything in the sub-app.
Another way could be that I use/modify `MultiAppProjectionTransfer` to mirror the mesh and project the data onto the new mesh, but I am not completely sure how to do it or if that is the best solution. Any clue?

Regards.

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think you are on the right track but I think the restart from exodus will be difficult. Like I dont know that we support restarting from two different exodus files. 
I would create two multiapps, flip one, restart both from exodus (or checkpoint), then do the transfers, each on their respective part of the full 3 D domain. 

Guillaume

**Reply by am-tc01** [reply]
Hi @GiudGiud 

Thanks! Yes I think that sounds more feasible. 
I could also create a full mesh using VTK (I simply flipped the original solution and grouped the data) but when I read it in Moose, I again got the error of `Negative Jacobian`. The question now is how to flip the solution and what exactly is causing this error?
Thanks!

**Reply by GiudGiud** [reply]
Flip the elements by using a rotation.
The negative jacobian is probably because the element ended up with a negative volume with the transform you applied.

**Reply by am-tc01** [reply]
> _Flip the elements by using a rotation._
How can I flip 3D geometry by using rotation? I tried `TransformGenerator `with `Scale = '1 1 -1',` but that resulted in the same error.

**Reply by GiudGiud** [reply]
Try ROTATE then TRANSLATE

**Reply by am-tc01** [reply]
Thanks! Have tried it already and that doesn't work.
Attached is the geometry I am trying to flip along XY plane.
[test1.zip](https://github.com/idaholab/moose/files/7989806/test1.zip)

**Reply by GiudGiud** [reply]
Hello

I just had a look and this seemed to stitch the meshes correctly (for 1/4 -> 1/2 of the geometry, might not have done the XY plane precisely)

```
[Mesh]
  [fmg_left]
    type = FileMeshGenerator
    file = test1.e
  []
  [fmg_right]
    type = FileMeshGenerator
    file = test1.e
  []
  [rotate]
    type = TransformGenerator
    input = 'fmg_right'
    transform = 'ROTATE'
    vector_value = '0 -90 0'
  []
  [translate]
    type = TransformGenerator
    input = 'rotate'
    transform = 'TRANSLATE'
    vector_value = '0 -0.001 0.001'
  []
  [smg]
    type = StitchedMeshGenerator
    inputs = 'fmg_left translate'
    clear_stitched_boundary_ids = true
    stitch_boundaries_pairs = 'bottom front'
    parallel_type = 'replicated'
  []
[]
```
Please let us know if that ends up working for your case

Guillaume

**Reply by am-tc01** [reply]
Hi @GiudGiud 
Thanks for having a look at it! So what you did now actually rotated the mesh, it didn't truly flip it (look at the temp data closely). I don't think one can flip the mesh by rotating+translating it. You have to change z = -z in order to flip it.
Anyhow, I have flipped the mesh now and fixed the error with negative Jacobian :) So now I have two files, original and flipped. The problem now is how to read a variable from both these files. I can either stitch them together and restart, but I guess Moose doesn't support that. Another way, like you said earlier, is to restart from two apps (one with original and other with flipped) and then do the transfer, each on their respective part on full 3D domain. Is that possible? I tried it but then one Transfer overwrites the other Transfer such that I have the variable data only from one Transfer. How can I combine the two? Thanks!
I am attaching the flipped mesh
[test1_flipped.zip](https://github.com/idaholab/moose/files/8024044/test1_flipped.zip)
.

**Reply by GiudGiud** [reply]
Well yes that is true. But dont you have an additional degree of symmetry in there? As in the system is 1/8th symmetry as far as I can see. 
If that is true, then even without flipping you can leverage that to reconstruct the full solution from the 1/4.

We dont have block restricted transfers yet. But you can create two additional (one additional would work actually, but let s do two for simplicity) additional auxiliary variables. Those will be block restricted to each symmetric part of the domain. You will first transfer the fields onto these, then use a ParsedAux or a SelfAux to copy the info from these variables onto the unified one.

**Reply by GiudGiud** [reply]
Ah really zooming in, it's not 1/8th symmetric.

**Reply by am-tc01** [reply]
> We dont have block restricted transfers yet. But you can create two additional (one additional would work actually, but let s do two for simplicity) additional auxiliary variables. Those will be block restricted to each symmetric part of the domain. You will first transfer the fields onto these, then use a ParsedAux or a SelfAux to copy the info from these variables onto the unified one.

Thanks for the tip! Yes something like that should work, but if I have two variables that are block restricted (one from each symmetric part), can I combine them using ParsedAux onto a unified variable (which is defined in both blocks)?

**Reply by GiudGiud** [reply]
you can be using two ParsedAux. 
The unified variable will be defined on both blocks, each ParsedAux will only be defined on one

**Reply by am-tc01** [reply]
OK, I didn't quite get that. I did the following but got an error.
I am using SolutionAux and SolutionUserObject to copy solution from original and flipped files.

```
[AuxVariables]
    [temp_original]
        block = '0'
    []
    [temp_flipped]
        block = '1'
    []
    [temp_full]
    []
[]

[AuxKernels]
    [initial_cond_original]
        type = SolutionAux
        solution = original_solution
        execute_on = initial
        variable = temp_original
    []
    [initial_cond_mirror]
        type = SolutionAux
        solution = mirror_solution
        execute_on = initial
        variable = temp_flipped
    []
    [initial_cond]
        type = ParsedAux
        variable = temp_full
        args = 'temp_original temp_flipped'
        function = 'temp_original + temp_flipped'
    []
[]
```

```
*** ERROR ***
The 'block' parameter of the object 'initial_cond' must be a subset of the 'block' parameter of the variable 'temp_original':
    Object 'initial_cond': 0, 1
    Variable 'temp_original': 0
```

**Reply by GiudGiud** [reply]
Try this:
```
    [initial_cond_1]
        type = ParsedAux
        variable = temp_full
        args = 'temp_flipped'
        function = 'temp_flipped'
        execute_on = initial
        block = 1
    []
    [initial_cond_2]
        type = ParsedAux
        variable = temp_full
        args = 'temp_original
        function = 'temp_original'
        execute_on = initial
        block = 0
    []
```

if this is for an initial condition, try to use an initial condition instead.
If there s no option that will let you parse this, or if the SolutionAux is not being execute pre-initial_conditions as expected, execute_on initial on an auxkernel is as close as you ll get

**Reply by am-tc01** [reply]
That worked! Thanks :)
